chain/ethereum/src/data_source.rs

```diff
@@ -358,13 +404,40 @@ impl DataSource {
     fn handler_for_block(
         &self,
         trigger_type: &EthereumBlockTriggerType,
+        block: BlockNumber,
     ) -> Option<MappingBlockHandler> {
         match trigger_type {
-            EthereumBlockTriggerType::Every => self
+            // Start matches only initialization handlers with a `once` filter
+            EthereumBlockTriggerType::Start => self
                 .mapping
                 .block_handlers
                 .iter()
-                .find(move |handler| handler.filter.is_none())
+                .find(move |handler| match handler.filter {
+                    Some(BlockHandlerFilter::Once) => block == self.start_block,
+                    _ => false,
+                })
+                .cloned(),
+            // End matches all handlers without a filter or with a `polling` filter
+            EthereumBlockTriggerType::End => self
+                .mapping
+                .block_handlers
+                .iter()
+                .find(move |handler| match handler.filter {
+                    Some(BlockHandlerFilter::Polling { every }) => {
+                        // POLLING_TODO: Get creation block if start_block is zero
+                        let start_block = self.start_block;
+                        let should_trigger = match every {
+                            // 0 is an invalid polling interval
+                            // A polling interval of 0 is used internally to indicate
+                            // that the block handler should be triggered once
+                            0 => false,
```

Comment:
Considering we now have the `Once` filter, we shouldn't need to set this to `0` for triggering only once. One cool thing you could try is using a [NonZeroU32](https://doc.rust-lang.org/stable/std/num/struct.NonZeroU32.html) to indicate that this is never zero.